### PR TITLE
Introduces the use of omitempty to the settings package

### LINKED ISSUE #60: Alow for default and/or required value/flag when requesting a key
A user of the settings package should when looking for keys be allowed to specify if the key requested is required.

A user of the settings package should when looking for a key be allowed to specify a default value that will be used as a fallback if no matching key is found.

### PR DESCRIPTION
**Pull request by GitCodeOli**
Allows a user to specify if a setting struct to be filleds fields can be ignored by using the omitempty tag.

The package will still return a list of missingkeys to the end user, and the values for these will be set to the default for that type.

MapFill has also been removed in this PR as it is redundant. 

Closes #60

### CODE REVIEW

File: filler.go [hunk 1]
```diff
@@ -73,25 +74,40 @@ func (pf *ProviderFiller) Fill(configPtr interface{}) error {
...
+		Omitempty: omit,
+	}
 }
 
 // fillStruct fills a struct, with the settings provided by the Provider
 // skipping invalid fields and fields it can't set.
 // fillStruct will error if a setting can't be loaded.
-func (pf *ProviderFiller) fillStruct(structVal reflect.Value, tags map[string]string) error {
+func (pf *ProviderFiller) fillStruct(structVal reflect.Value, keysPacks map[string]keyPackage) error {
+	tags, requiredKeys := determieRequiredTags(keysPacks)
```
Comment (alessandrozucca) [inline filler.go]: typo `determie`

File: filler.go [hunk 2]
```diff
@@ -73,25 +74,40 @@ func (pf *ProviderFiller) Fill(configPtr interface{}) error {
...
+
 	settings, err := fetchValuesForTags(pf.Provider, tags)
 	if err != nil {
-		return fmt.Errorf("%v are not valid settings (%v)", err.MissingKeys(), err.Error())
+		// bypassing error if all the missingKeys are tagged omitempty
+		// otherwise we return the full list of missingkeys (tagged omitempty or not) - with an error
+		for _, v := range requiredKeys {
+			if stringInSlice(v, err.MissingKeys()) {
```
Comment (alessandrozucca) [inline filler.go]: we could move `err.MissingKeys()` before the loop so is only done once

File: filler.go [hunk 3]
```diff
@@ -73,25 +74,40 @@ func (pf *ProviderFiller) Fill(configPtr interface{}) error {
...
-func (pf *ProviderFiller) fillStruct(structVal reflect.Value, tags map[string]string) error {
+func (pf *ProviderFiller) fillStruct(structVal reflect.Value, keysPacks map[string]keyPackage) error {
+	tags, requiredKeys := determieRequiredTags(keysPacks)
+
 	settings, err := fetchValuesForTags(pf.Provider, tags)
 	if err != nil {
-		return fmt.Errorf("%v are not valid settings (%v)", err.MissingKeys(), err.Error())
+		// bypassing error if all the missingKeys are tagged omitempty
+		// otherwise we return the full list of missingkeys (tagged omitempty or not) - with an error
+		for _, v := range requiredKeys {
```
Comment (alessandrozucca) [inline filler.go]: not sure if is worth it but if `determineRequiredTags` returns a `map[string]`with the `requiredKeys` you could invert the logic
```
pseudo code:
for missingKey in err.MissingKeys() {
  if requiredKeys[missingKey]  {
    return
  }
}
```

File: filler.go (hunk 2 above)
Comment (GitCodeOli) [reply filler.go]: Done

File: filler.go (hunk 3 above)
Comment (GitCodeOli) [reply filler.go]: I suppose we hope there are less missing keys than required ones - so i've done this.

File: filler.go [hunk 4]
```diff
@@ -73,25 +74,41 @@ func (pf *ProviderFiller) Fill(configPtr interface{}) error {
...
-// MapFill takes a pointer to a struct map of field names to setting keys
-// if the the field name does not exist then it will be ignored, if the
-// setting key does not exist then it will error.
-func (pf *ProviderFiller) MapFill(configPtr interface{}, tags map[string]string) error {
-	structVal, err := fetchValidStructValueFromPointer(configPtr)
-	if err != nil {
-		return err
-	}
+// KeyPackage contains the searched for key and whether this key is required or not
+type keyPackage struct {
```
[the remark this answers is not in the document]
Comment (GitCodeOli) [reply filler.go]: on it

File: Gopkg.toml [hunk 5]
```diff
@@ -6,3 +6,7 @@
 [[constraint]]
   name = "github.com/intelligentpos/structextract"
   version = "1.1.0"
+
+[[constraint]]
+  name = "github.com/stretchr/testify"
```
Comment (geototti21) [inline Gopkg.toml]: where this package is used? can we please dont use it and use the stlib  testing package!

File: filler.go [hunk 6]
```diff
@@ -73,25 +74,41 @@ func (pf *ProviderFiller) Fill(configPtr interface{}) error {
 	return pf.fillStruct(structVal, pf.fieldNamesToTags(structVal.Type()))
 }
 
-// MapFill takes a pointer to a struct map of field names to setting keys
-// if the the field name does not exist then it will be ignored, if the
-// setting key does not exist then it will error.
-func (pf *ProviderFiller) MapFill(configPtr interface{}, tags map[string]string) error {
```
Comment (geototti21) [inline filler.go]: Can we also update readme file when we are doing these changes, its a public repo.

Comment (GitCodeOli) [reply filler.go]: Removed and left a note

File: Gopkg.toml (hunk 5 above)
Comment (alessandrozucca) [reply Gopkg.toml]: I suggested him to use it, keep the assertion clean and in 1 single line. The standard `testing` package doesn't support assertions so you need to write them by hand

Comment (GitCodeOli) [reply Gopkg.toml]: it's used in default_provider_test.go for asserts. It's used in other parts of IPOS code base, is it not ok to use?

Comment (geototti21) [reply Gopkg.toml]: I usually avoid it but @alessandrozucca is in ❤️ with it. its fine then

**Pull request merged**